File: solutions/Leetcode_778/leetcode_778.cc
#include <queue>
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class Solution1
{
public:
    int swimInWater(vector<vector<int>> &grid)
    {
        unordered_map<int, queue<pair<int, int>>> time;
        time[grid[0][0]].push({0, 0});
        unordered_set<pair<int, int>> visited{{0, 0}};
        int n = grid.size();
        int directions[5]{1, 0, -1, 0, 1};
        for (int t = 0; t < n; ++t)
        {
            while (time.find(t) != time.end() && !time[t].empty())
            {
                int x = time[t].front().first, y = time[t].front().second;
                time[t].pop();
                if (x == n - 1 && y == n - 1)
                    return t;
                for (int i; i < 4; ++i)
                {
                    int xx = x + directions[i], yy = y + directions[i + 1];
                    if (0 <= xx && xx < n && 0 <= yy && yy < n && visited.find(pair<int, int>{xx, yy}) == visited.end())
                    {
                        visited.insert({xx, yy});
                        time[max(grid[xx][yy], t)].push({xx, yy});
                    }
                }
            }
        }
    }
};

File: solutions/Leetcode_778/leetcode_778.py
from collections import defaultdict
from typing import List


class Solution1:
    def swim_in_water(self, grid: List[List[int]]) -> int:
        time = defaultdict(list)
        time[grid[0][0]].append((0, 0))
        visited = set()
        n = len(grid)
        for t in range(n**2):
            while time[t]:
                x, y = time[t].pop()
                if x == y == n - 1:
                    return t
                for dx, dy in ((1, 0), (0, 1), (-1, 0), (0, -1)):
                    if (0 <= x+dx < n
                            and 0 <= y+dy < n
                            and (x+dx, y+dy) not in visited):
                        visited.add((x+dx, y+dy))
                        time[max(grid[x+dx][y+dy], t)].append((x+dx, y+dy))
